lib/metasearch_index/deleter.rb

```diff
@@ -0,0 +1,55 @@
+module MetasearchIndex
+  module Deleter
+    class V2
+      class MissingArgument < StandardError; end
+      class NotFound < StandardError; end
+      class UnknownError < StandardError; end
+
+      def initialize(id:, data:)
+        @id = id.presence || raise(MissingArgument, "ID and Type must be supplied.")
+        @data = data.presence || raise(MissingArgument, "Not record provided to insert into Elasticsearch.")
+      end
+
+      def delete
+        processor = ElasticsearchProcessor.new
+        processor.delete(self)
+        response = processor.commit
+        validate_response!(response.first)
+      end
+
+      def identifier
+        {
+          _type: _type,
+          _id: base_path,
+        }
+      end
+
+      def validate_response!(response)
+        action_type, details = response.first # response is a hash with a single [key, value] pair
+        status = details['status']
+
+        if (200..399).cover?(status)
+          logger.debug("Processed #{action_type} with status #{status}")
+          Services.statsd_client.increment("metasearch.elasticsearch.#{action_type}")
+        elsif status == 404 # failed while attempting to delete missing record so just ignore it
+          logger.info("Tried to delete a document that wasn't there; ignoring.")
+          Services.statsd_client.increment('metasearch.elasticsearch.already_deleted')
+          raise NotFound, "Document not found in index"
+        else
+          logger.error("#{action_type} not processed: status #{status}")
+          Services.statsd_client.increment("metasearch.elasticsearch.#{action_type}_error")
+
+          # manually send the error as we rescue it later in the process
```

Comment:
Maybe we could send this at the point we rescue the exception, as whether we still want to notify seems related to the context its being caught in

Reply:
yeah, I might even be able to drop this as need to look at what sinatra actually does in error mode.